### PR TITLE
Line plots with 2D coordinates

### LINKED ISSUE #2407: plot.line does not accept non-dimension coordinates
plot.line does not accept non-dimension coordinates
(https://stackoverflow.com/questions/52103604/xarray-plot-line-with-non-dimensional-coordinate/52190961#52190961)

```python
import xarray as xr
import numpy as np
import matplotlib.pyplot as plt

# Setup data array, from http://xarray.pydata.org/en/stable/plotting.html#multidimensional-coordinates
lon, lat = np.meshgrid(np.linspace(-20, 20, 5), 
                       np.linspace(0, 30, 4))
lon += lat/10
lat += lon/10
da = xr.DataArray(np.arange(20).reshape(4, 5), dims=['y','x'],
                     coords = {'lat': (('y', 'x'), lat),
                               'lon': (('y', 'x'), lon)})

# plot in terms of y,x, the dimensional coordinate
fig,ax = plt.subplots()
da.plot.pcolormesh('y','x') 

# plot in terms of lon, lat, the non-dimensional coordinate
fig, ax = plt.subplots()
da.plot.pcolormesh('lon', 'lat')

# plot lines in terms of x,y, the dimensional coordinate
fig, ax = plt.subplots()
da.plot.line(x='x')

# plot lines in terms of lon, lat?
da.plot.line(x='lon') #gives error

```
#### Problem description

It would be nice if I could plot line cuts through 2D data with a non-dimensional coordinate.  It is possible with 2D plotting.  

#### Output of ``xr.show_versions()``
```
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 3.19.3-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.8
pandas: 0.23.3
numpy: 1.15.0
scipy: 1.1.0
netCDF4: 1.4.0
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: None
bottleneck: None
cyordereddict: None
dask: None
distributed: None
matplotlib: 2.2.2
cartopy: None
seaborn: 0.9.0
setuptools: 40.0.0
pip: 18.0
conda: None
pytest: None
IPython: 6.4.0
sphinx: None

```

<details>
# Paste the output here xr.show_versions() here

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #2407 
 - [x] Tests added
 - [x] Tests passed 
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by dcherian** [comment]
This is ready for review. The API is `da.plot.line(x='lon', hue='x')`. `hue` must be a dimension name in this case.

File: doc/plotting.rst [hunk 1]
```diff
@@ -702,3 +702,12 @@ You can however decide to infer the cell boundaries and use the
     outside the xarray framework.
 
 .. _cell boundaries: http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#cell-boundaries
+
+One can also make line plots with multidimensional co-ordinates. In this case, ``hue`` must be a dimension name, not a co-ordinate name.
```
Comment (fmaussion) [inline doc/plotting.rst]: Nit: I don't think you need an hyphen in coordinates. We write it without hyphen in the [data structures](http://xarray.pydata.org/en/latest/data-structures.html#coordinates) section.

File: xarray/plot/plot.py [hunk 2]
```diff
@@ -221,18 +222,38 @@ def _infer_line_data(darray, x, y, hue):
...
-            yplt = darray.transpose(xname, huename)
+            xplt = darray[xname]
+            if xplt.ndim > 1:
+                if huename in darray.dims:
+                    otherdim = darray.dims[
+                        np.mod(darray.dims.index(huename) + 1, 2)]
+                    yplt = darray.transpose(otherdim, huename)
+                    xplt = xplt.transpose(otherdim, huename)
+                else:
+                    raise ValueError('If x or y are 2D, hue must be a dimension' +
```
Comment (fmaussion) [inline xarray/plot/plot.py]: -> ``hue must be a dimension name '`` (mind the space character at the end of lines, also line  below)

File: xarray/plot/plot.py [hunk 3]
```diff
@@ -221,18 +222,38 @@ def _infer_line_data(darray, x, y, hue):
...
         else:
             yname, huename = _infer_xy_labels(darray=darray, x=y, y=hue)
-            xname = darray.name
-            xplt = darray.transpose(yname, huename)
-            yplt = darray.coords[yname]
+            yplt = darray[yname]
+            if yplt.ndim > 1:
+                if huename in darray.dims:
+                    otherdim = darray.dims[
+                        np.mod(darray.dims.index(huename) + 1, 2)]
```
Comment (fmaussion) [inline xarray/plot/plot.py]: I'm not sure to understand the call to ``np.mod`` here. This is going to work with 2D arrays only, right?

File: xarray/plot/plot.py [hunk 4]
```diff
@@ -265,9 +286,11 @@ def line(darray, *args, **kwargs):
         Axis on which to plot this figure. By default, use the current axis.
         Mutually exclusive with ``size`` and ``figsize``.
     hue : string, optional
-        Coordinate for which you want multiple lines plotted.
+        Dimension or coordinate for which you want multiple lines plotted.
+        If plotting against a 2D co-ordinate, ``hue`` must be a dimension.
```
Comment (fmaussion) [inline xarray/plot/plot.py]: hyphen

File: xarray/tests/test_plot.py [hunk 5]
```diff
@@ -1806,3 +1805,23 @@ def test_yticks_kwarg(self, da):
         da.plot(yticks=np.arange(5))
         expected = np.arange(5)
         assert np.all(plt.gca().get_yticks() == expected)
+
+
+def test_multidim_coords_line_plot():
```
Comment (fmaussion) [inline xarray/tests/test_plot.py]: This test should be added to a Unittest child class -> where are the other line plots tested?

Comment (dcherian) [reply xarray/tests/test_plot.py]: Oh it needed a `@requires_matplotlib`. I moved it up but I thought we were supposed to be transitioning to a classless testing code.

File: xarray/plot/plot.py (hunk 3 above)
Comment (dcherian) [reply xarray/plot/plot.py]: yup, addressed.

**Comment by dcherian** [comment]
dask-dev test timed out for some reason but others have passed.

File: xarray/plot/plot.py (hunk 3 above)
Comment (fmaussion) [reply xarray/plot/plot.py]: This should be the same as in the xcase (you changed only the block above).

Note that the same could be achieved with ``otherdim = darray.dims[int(darray.dims.index(huename) == 0)]`` (maybe less explicit).

File: doc/whats-new.rst [hunk 6]
```diff
@@ -33,6 +33,10 @@ v0.11.0 (unreleased)
 Enhancements
 ~~~~~~~~~~~~
 
+- ``plot.line()`` can now accept multidimensional co-ordinate variables as input.
```
Comment (fmaussion) [inline doc/whats-new.rst]: hyphen

**Pull request merged**